File: src/array/WordSearch_79.java
package array;

public class WordSearch_79 {

    public static void main(String[] args) {
        char[][] board = {
                {'A', 'B', 'C', 'E'},
                {'S', 'F', 'C', 'S'},
                {'A', 'D', 'E', 'E'}};
        System.out.println(new WordSearch_79().exist(board, "ABCCED"));
        System.out.println(new WordSearch_79().exist(board, "SEE"));
        System.out.println(new WordSearch_79().exist(board, "ABCB"));
    }

    public boolean exist(char[][] board, String word) {

        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == word.charAt(0)) {
                    if (exist(board, i, j, word.toCharArray(), 0))
                        return true;
                }
            }
        }
        return false;
    }

    private boolean exist(char[][] board, int i, int j, char[] word, int index) {
        if (i < 0 || j < 0 || i > board.length - 1 || j > board[0].length - 1)
            return false;
        if (board[i][j] == '*')
            return false;
        if (board[i][j] != word[index])
            return false;

        if (word.length - 1 == index)
            return true;

        board[i][j] = '*';

        //search

        boolean result = exist(board, i - 1, j, word, index + 1)
                || exist(board, i, j + 1, word, index + 1)
                || exist(board, i + 1, j, word, index + 1)
                || exist(board, i, j - 1, word, index + 1);

        board[i][j] = word[index];

        return result;
    }
}

File: src/backtracking/word_search_79.py
# loop all board
# find first char
# dfs the rest of them
# store visited
# return true if found the last one

class Solution(object):
    def exist(self, board, word):
        n = len(board)
        m = len(board[0])

        visited = [[False for _ in range(m)] for _ in range(n)]

        for i in range(n):
            for j in range(m):
                if board[i][j] == word[0]:
                    if self.dfs(board, word, i, j, 0, visited):
                        return True

        return False

    def dfs(self, board, word, i, j, index, visited):
        if i >= len(board) or i < 0 or j >= len(board[0]) or j < 0:
            return False

        if board[i][j] != word[index] or visited[i][j] :
            return False

        if index == len(word) - 1:
            return True

        visited[i][j] = True

        if self.dfs(board, word, i - 1, j, index + 1, visited) or \
                self.dfs(board, word, i, j + 1, index + 1, visited) or \
                self.dfs(board, word, i + 1, j, index + 1, visited) or \
                self.dfs(board, word, i, j - 1, index + 1, visited):
            return True

        visited[i][j] = False

        return False


print(Solution().exist(board=[["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], word="ABCCED"))
print(Solution().exist(board=[["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], word="SEE"))
print(Solution().exist(board=[["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], word="ABCB"))
